File: js-template/leetcode/editor/cn/set-matrix-zeroes.js
/*
 * @lc app=leetcode.cn id=73 lang=javascript
 * @lcpr version=30404
 *
 * [73] 矩阵置零
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var setZeroes = function(matrix) {
    var row = new Set();
    var col = new Set();
    var m = matrix.length;
    var n = matrix[0].length;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (matrix[i][j] === 0) {
                row.add(i);
                col.add(j);
            }
        }
    }

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (row.has(i) || col.has(j))
                matrix[i][j] = 0;
        }
    }
};
// @lc code=end

// your test code here





/*
// @lcpr case=start
// [[1,1,1],[1,0,1],[1,1,1]]\n
// @lcpr case=end

// @lcpr case=start
// [[0,1,2,0],[3,4,5,2],[1,3,1,5]]\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/set-matrix-zeroes.cpp
/*
 * @lc app=leetcode.cn id=73 lang=cpp
 * @lcpr version=30404
 *
 * [73] 矩阵置零
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        unordered_set<int> rows, cols;
        int m = matrix.size();
        int n = matrix[0].size();
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] == 0) {
                    rows.insert(i);
                    cols.insert(j);
                }
            }
        }
        for (auto &&row : rows) {
            for (int i = 0; i < n; ++i) {
                matrix[row][i] = 0;
            }
        }
        for (auto &&col : cols) {
            for (int i = 0; i < m; ++i)
                matrix[i][col] = 0;
        }
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
}



/*
// @lcpr case=start
// [[1,1,1],[1,0,1],[1,1,1]]\n
// @lcpr case=end

// @lcpr case=start
// [[0,1,2,0],[3,4,5,2],[1,3,1,5]]\n
// @lcpr case=end

 */

